third_party/xla_client/pjrt_computation_client.cc

```diff
@@ -278,8 +298,8 @@ PjRtComputationClient::ExecuteComputation(
     std::unique_ptr<xla::PjRtBuffer> buffer = std::move(result);
 
     std::shared_ptr<PjRtData> data = std::make_shared<PjRtData>(
-        device, buffer->logical_on_device_shape().ValueOrDie(),
-        std::move(buffer));
+        // TODO(wcromar): do we need `logical_on_device_shape` here?
+        device, buffer->on_device_shape(), std::move(buffer));
```

Comment:
@JackCaoG Do you know if we actually need to use `logical_on_device_shape` here instead of `on_device_shape`? Here's the difference from the [doc](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/pjrt_client.h#L751-L757):

```
  // Same as on_device_shape when the shape is static. When the shape is
  // dynamic, it gathers the metadata from the device and returns a static shape
  // representing the logical shape of the data. This approach is identical to
  // how tensorflow and xrt setup the output buffer in the graph.
  //
  // Since this method actually acquires locks and communicate with the device,
  // it does not have the const qualifier, similar to what ToLiteral does.
  ```
  
  To play it safe and avoid breaking SPMD, I left `logical_on_device_shape` in `ExecuteReplicated`. If we need `logical_on_device_shape`, I can file a feature request to PJRT to implement this in the C API.

Reply:
From the comments it seems like we need `logical_on_device_shape` for dynamic shape tensors. We should file a feature request. 

The more concerning part of this thing is when I look up the tfrt implementation of this function I found
```
Dynamic shapes are not implemented in the TFRT backend
```

I think we need to identify the gap between streamexecutor and tfrt regarding the dynamic shape support. @miladm

Reply:
For this PR, I modified this line to only use `on_device_shape` when we're using TFRT. Otherwise, we'll continue to use `logical_on_device_shape`.

I also filed a feature request to implement `logical_on_device_shape` in the C API, but the underlying TFRT TPU implementation of this function might not support dynamic shapes.

Reply:
Sounds like TFRT does not adopt DS yet; we should certainly work on landing it as part of PjRt adoption. How do we start this line of discussion? @will-cromar @JackCaoG

Reply:
FYI Regarding SPMD, the output shape will always be static. So using `on_device_shape` won't break. It does sound like we need the support, though, for dynamic shape.